IFrameFullscreenFlag doesn't need to be on Node

rniwa opened the pull request:
#### 519d5e9d71a1a8e89e96dd82797544323fdee474
<pre>
IFrameFullscreenFlag doesn&apos;t need to be on Node
<a href="https://bugs.webkit.org/show_bug.cgi?id=266812">https://bugs.webkit.org/show_bug.cgi?id=266812</a>

Reviewed by Tim Nguyen and Alan Baradlay.

Move IFrameFullscreenFlag from Node to HTMLIFrameElement where this flag is only ever set true.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setFullscreenFlag):
(WebCore::Element::setIFrameFullscreenFlag): Deleted.
* Source/WebCore/dom/Element.h:
(WebCore::Element::hasFullscreenFlag const):
(WebCore::Element::hasIFrameFullscreenFlag const): Deleted.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::documentsToUnfullscreen):
(WebCore::clearFullscreenFlags): Added.
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::finishExitFullscreen):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::exitRemovedFullscreenElement):
* Source/WebCore/dom/Node.h:
* Source/WebCore/html/HTMLIFrameElement.h:

Canonical link: <a href="https://commits.webkit.org/272459@main">https://commits.webkit.org/272459@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af8596270cf6e0e094caa04480e7963ef1f04b0e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31704 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33427 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34196 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28709 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32488 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7635 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28304 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8754 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28296 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7554 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7714 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28210 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35541 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28818 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28657 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33837 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7808 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5813 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31693 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9463 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8487 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4143 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8324 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -537,6 +546,8 @@ bool FullscreenManager::willEnterFullscreen(Element& element)
 
         ancestor->setFullscreenFlag(true);
+        if (auto* frameElement = dynamicDowncast<HTMLIFrameElement>(ancestor.get()))
+            frameElement->setIFrameFullscreenFlag(false);
```
I can't remember why I set the flag to false in this specific case, it's not in the spec: https://fullscreen.spec.whatwg.org/#dom-element-requestfullscreen

Might be possible to remove it.

rniwa replied:
I'd add a FIXME.

nt1m commented on Source/WebCore/html/HTMLIFrameElement.h:
```diff
@@ -70,6 +75,7 @@ class HTMLIFrameElement final : public HTMLFrameElementBase {
 
     std::unique_ptr<DOMTokenList> m_sandbox;
     mutable std::optional<FeaturePolicy> m_featurePolicy;
+    bool m_iframeFullsccreenFlag { false };
```
You could gate this behind `ENABLE(FULLSCREEN_API)` too fwiw.

nt1m replied:
Also there are 2 `c` in `Fullsccreen`

rniwa committed:
IFrameFullscreenFlag doesn't need to be on Node
https://bugs.webkit.org/show_bug.cgi?id=266812

Reviewed by Tim Nguyen and Alan Baradlay.

Move IFrameFullscreenFlag from Node to HTMLIFrameElement where this flag is only ever set true.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setFullscreenFlag):
(WebCore::Element::setIFrameFullscreenFlag): Deleted.
* Source/WebCore/dom/Element.h:
(WebCore::Element::hasFullscreenFlag const):
(WebCore::Element::hasIFrameFullscreenFlag const): Deleted.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::documentsToUnfullscreen):
(WebCore::clearFullscreenFlags): Added.
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::finishExitFullscreen):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::exitRemovedFullscreenElement):
* Source/WebCore/dom/Node.h:
* Source/WebCore/html/HTMLIFrameElement.h:

Canonical link: https://commits.webkit.org/272459@main
